fs/io_zenfs.cc

```diff
@@ -415,8 +420,65 @@ IOStatus ZoneFile::AllocateNewZone() {
   return PersistMetadata();
 }
 
+/* Byte-aligned, sparse writes with inline metadata
+   the caller reserves 8 bytes of data for a size header */
+IOStatus ZoneFile::SparseAppend(char* sparse_buffer, uint32_t data_size) {
+  uint32_t left = data_size;
+  uint32_t wr_size;
+  uint32_t block_sz = GetBlockSize();
+  IOStatus s;
+
+  if (active_zone_ == NULL) {
+    s = AllocateNewZone();
+    if (!s.ok()) return s;
+  }
+
+  while (left) {
+    wr_size = left + ZoneFile::SPARSE_HEADER_SIZE;
+    if (wr_size > active_zone_->capacity_) wr_size = active_zone_->capacity_;
+
+    /* Pad to the next block boundary if needed */
+    uint32_t align = wr_size % block_sz;
+    uint32_t pad_sz = 0;
+
+    align = wr_size % block_sz;
+    if (align) pad_sz = block_sz - align;
+    if (pad_sz) memset(sparse_buffer + wr_size, 0x0, pad_sz);
+
+    uint64_t extent_length = wr_size - ZoneFile::SPARSE_HEADER_SIZE;
+    EncodeFixed64(sparse_buffer, extent_length);
+
+    s = active_zone_->Append(sparse_buffer, wr_size + pad_sz);
+    if (!s.ok()) return s;
+
+    extents_.push_back(
+        new ZoneExtent(extent_start_ + ZoneFile::SPARSE_HEADER_SIZE,
+                       extent_length, active_zone_));
+
+    extent_start_ = active_zone_->wp_;
+    active_zone_->used_capacity_ += extent_length;
+    fileSize += extent_length;
+    left -= extent_length;
+
+    if (active_zone_->capacity_ == 0) {
+      s = CloseActiveZone();
+      if (!s.ok()) {
+        return s;
+      }
+      if (left) {
+        memcpy((void*)(sparse_buffer + ZoneFile::SPARSE_HEADER_SIZE),
+               (void*)(sparse_buffer + wr_size), left);
+      }
```

Comment:
Why not move the pointer instead of the data?

Reply:
Because of alignment reasons, The data has to start on a page boundary to issue direct io.